File: inventra/lib/core/utils/validators.dart
class Validators {
  static String? email(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Email is required';
    }
    final emailRegex = RegExp(r'^[^\s@]+@[^\s@]+\.[^\s@]+$');
    if (!emailRegex.hasMatch(value.trim())) {
      return 'Enter a valid email';
    }
    return null;
  }

  static String? password(String? value) {
    if (value == null || value.isEmpty) {
      return 'Password is required';
    }
    if (value.length < 8) {
      return 'At least 8 characters';
    }
    if (!RegExp(r'[A-Z]').hasMatch(value)) {
      return 'Include an uppercase letter';
    }
    if (!RegExp(r'[a-z]').hasMatch(value)) {
      return 'Include a lowercase letter';
    }
    if (!RegExp(r'\d').hasMatch(value)) {
      return 'Include a number';
    }
    return null;
  }

  static String? fullName(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Full name is required';
    }
    if (value.trim().length < 3) {
      return 'At least 3 characters';
    }
    return null;
  }

  static String? required(String? value, {String field = 'This field'}) {
    if (value == null || value.trim().isEmpty) {
      return '$field is required';
    }
    return null;
  }

  static String? positiveNumber(String? value, {String field = 'Value'}) {
    if (value == null || value.trim().isEmpty) {
      return '$field is required';
    }
    final number = double.tryParse(value);
    if (number == null || number < 0) {
      return 'Enter a valid number';
    }
    return null;
  }
}

File: backend/app/utils/validators.py
import re


def validate_registration(data):
    errors = {}

    full_name = data.get("full_name", "").strip()
    email = data.get("email", "").strip()
    password = data.get("password", "")

    # Full Name
    if not full_name:
        errors["full_name"] = "Full name is required"
    elif len(full_name) < 3:
        errors["full_name"] = "Full name must be at least 3 characters"
    elif len(full_name) > 100:
        errors["full_name"] = "Full name cannot exceed 100 characters"

    # Email
    email_regex = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"

    if not email:
        errors["email"] = "Email is required"
    elif not re.match(email_regex, email):
        errors["email"] = "Invalid email format"

    # Password
    if not password:
        errors["password"] = "Password is required"
    elif len(password) < 8:
        errors["password"] = "Password must be at least 8 characters"
    elif not re.search(r"[A-Z]", password):
        errors["password"] = "Password must contain at least one uppercase letter"
    elif not re.search(r"[a-z]", password):
        errors["password"] = "Password must contain at least one lowercase letter"
    elif not re.search(r"\d", password):
        errors["password"] = "Password must contain at least one number"

    return errors